src/lib/statistics/table_statistics.cpp

```diff
@@ -1,317 +1,96 @@
 
-  // Estimate "a BETWEEN 5 and 6" by combining "a >= 5" with "a <= 6"
-  if (is_between_predicate_condition(predicate_condition)) {
-    DebugAssert(value2, "Expected second value to be passed in for BETWEEN");
-    auto table_statistics = estimate_predicate(column_id, PredicateCondition::GreaterThanEquals, value);
-    return table_statistics.estimate_predicate(column_id, PredicateCondition::LessThanEquals, *value2);
-  }
-
-  // TODO(anybody) we don't do (Not)Like estimations yet, thus resort to magic numbers
-  if (predicate_condition == PredicateCondition::Like || predicate_condition == PredicateCondition::NotLike) {
-    const auto selectivity =
-        predicate_condition == PredicateCondition::Like ? DEFAULT_LIKE_SELECTIVITY : 1.0f - DEFAULT_LIKE_SELECTIVITY;
-    return {TableType::References, _row_count * selectivity, _column_statistics};
-  }
+          auto histogram = std::shared_ptr<AbstractHistogram<ColumnDataType>>{};
 
-  // Create copies to modify below and insert into result
-  auto predicated_row_count = _row_count;
-  auto predicated_column_statistics = _column_statistics;
+          histogram =
+              EqualDistinctCountHistogram<ColumnDataType>::from_column(table, my_column_id, histogram_bin_count);
 
-  const auto left_operand_column_statistics = _column_statistics[column_id];
+          if (histogram) {
+            output_column_statistics->set_statistics_object(histogram);
 
-  if (predicate_condition == PredicateCondition::IsNotNull) {
-    predicated_column_statistics[column_id] = left_operand_column_statistics->without_null_values();
-    predicated_row_count *= 1.0f - left_operand_column_statistics->non_null_value_ratio();
-  } else if (predicate_condition == PredicateCondition::IsNull) {
-    predicated_column_statistics[column_id] = left_operand_column_statistics->only_null_values();
-    predicated_row_count *= left_operand_column_statistics->non_null_value_ratio();
-  } else if (is_column_id(value)) {
-    const auto column_id_of_value = boost::get<ColumnID>(value);
-    const auto estimation = left_operand_column_statistics->estimate_predicate_with_column(
-        predicate_condition, *_column_statistics[column_id_of_value]);
+            // Use the insight the the histogram will only contain non-null values to generate the NullValueRatio
+            // property
+            const auto null_value_ratio =
+                table.row_count() == 0 ? 0.0f
+                                       : 1.0f - (static_cast<float>(histogram->total_count()) / table.row_count());
+            output_column_statistics->set_statistics_object(
+                std::make_shared<NullValueRatioStatistics>(null_value_ratio));
+          } else {
+            // Failure to generate a histogram currently only stems from all-null segments.
+            // TODO(anybody) this is a slippery assumption. But the alternative would be a full segment scan...
```

Comment:
At least check it? With a 100% null segment, someone else might introduce pruning rules for something that is apparently empty.

Reply:
As the comment states "But the alternative would be a full segment scan". This Assert would be pretty awkward to type.

People need to be very careful with pruning ;)